plugins/fs/src/commands.rs

```diff
@@ -40,360 +56,747 @@ impl Serialize for CommandError {
+    to_path_base_dir: Option<BaseDirectory>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct MkdirOptions {
+    #[serde(flatten)]
+    base: BaseOptions,
+    #[allow(unused)]
+    mode: Option<u32>,
+    recursive: Option<bool>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct RemoveOptions {
+    #[serde(flatten)]
+    base: BaseOptions,
+    recursive: Option<bool>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct RenameOptions {
+    new_path_base_dir: Option<BaseDirectory>,
+    old_path_base_dir: Option<BaseDirectory>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct WriteFileOptions {
+    #[serde(flatten)]
+    base: BaseOptions,
+    append: Option<bool>,
+    create: Option<bool>,
+    #[allow(unused)]
+    mode: Option<u32>,
+}
+
+#[derive(Serialize_repr, Deserialize_repr, Clone, Copy, Debug)]
+#[repr(u16)]
+pub enum SeekMode {
+    Start = 0,
+    Current = 1,
+    End = 2,
 }
 
 #[tauri::command]
-pub fn read_file<R: Runtime>(
-    window: Window<R>,
+pub fn create<R: Runtime>(
+    app: AppHandle<R>,
     path: SafePathBuf,
-    options: Option<FileOperationOptions>,
-) -> CommandResult<Vec<u8>> {
-    let resolved_path = resolve_path(&window, path, options.and_then(|o| o.dir))?;
-    fs::read(&resolved_path)
-        .with_context(|| format!("path: {}", resolved_path.display()))
-        .map_err(Into::into)
+    options: Option<BaseOptions>,
+) -> CommandResult<ResourceId> {
+    let resolved_path = resolve_path(&app, path, options.and_then(|o| o.dir))?;
+    let file = File::create(&resolved_path).map_err(|e| {
+        format!(
+            "failed to create file at path: {} with error: {e}",
+            resolved_path.display()
+        )
+    })?;
+    let rid = app.resources_table().add(StdFileResource::new(file));
+    Ok(rid)
 }
 
 #[tauri::command]
-pub fn read_text_file<R: Runtime>(
-    window: Window<R>,
+pub fn open<R: Runtime>(
+    app: AppHandle<R>,
     path: SafePathBuf,
-    options: Option<FileOperationOptions>,
-) -> CommandResult<String> {
-    let resolved_path = resolve_path(&window, path, options.and_then(|o| o.dir))?;
-    fs::read_to_string(&resolved_path)
-        .with_context(|| format!("path: {}", resolved_path.display()))
-        .map_err(Into::into)
+    options: Option<OpenOptions>,
+) -> CommandResult<ResourceId> {
+    let resolved_path = resolve_path(&app, path, options.as_ref().and_then(|o| o.base.dir))?;
+
+    let mut opts = std::fs::OpenOptions::new();
+
+    match options {
+        None => {
+            opts.read(true)
+                .create(false)
+                .write(false)
+                .truncate(false)
+                .append(false)
+                .create_new(false);
+        }
+        Some(options) => {
+            #[cfg(unix)]
+            {
+                use std::os::unix::fs::OpenOptionsExt;
+                if let Some(mode) = options.mode {
+                    opts.mode(mode & 0o777);
```

Comment:
shouldn't we just set the mode here?

Reply:
tbh I don't know, this was copied from deno implementation but I guess they have a default mode or something, I will just change that to use the mode directly.